Source/GSStoryboardTransform.m

```diff
@@ -0,0 +1,828 @@
+/* Implementation of class GSStoryboardTransform
+   Copyright (C) 2020 Free Software Foundation, Inc.
+   
+   By: Gregory John Casamento
+   Date: Sat 04 Jul 2020 03:48:15 PM EDT
+
+   This file is part of the GNUstep Library.
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import <Foundation/NSData.h>
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSString.h>
+#import <Foundation/NSMapTable.h>
+#import <Foundation/NSXMLDocument.h>
+#import <Foundation/NSXMLNode.h>
+#import <Foundation/NSXMLElement.h>
+#import <Foundation/NSUUID.h>
+#import <Foundation/NSBundle.h>
+#import <Foundation/NSKeyedArchiver.h>
+#import <Foundation/NSException.h>
+
+#import "AppKit/NSSeguePerforming.h"
+#import "AppKit/NSStoryboard.h"
+#import "AppKit/NSStoryboardSegue.h"
+#import "AppKit/NSNibDeclarations.h"
+#import "AppKit/NSViewController.h"
+#import "AppKit/NSWindowController.h"
+
+#import "GSStoryboardTransform.h"
+#import "GSFastEnumeration.h"
+
+#define APPLICATION @"application"
+
+@interface NSStoryboardSegue (__private__)
+- (void) _setDestinationController: (id)controller; 
+- (void) _setSourceController: (id)controller;
+@end
+
+@interface NSStoryboardSegue (__StoryboardPrivate__)
+// Private to this class...
+- (void) _setKind: (NSString *)k;
+- (void) _setRelationship: (NSString *)r;
+- (NSString *) _kind;
+- (NSString *) _relationship;
+@end
+
+// this needs to be set on segues
+@implementation NSStoryboardSegue (__StoryboardPrivate__)
+- (void) _setKind: (NSString *)k
+{
+  ASSIGN(_kind, k);
+}
+
+- (void) _setRelationship: (NSString *)r
+{
+  ASSIGN(_relationship, r);
+}
+
+- (NSString *) _kind
+{
+  return _kind;
+}
+
+- (NSString *) _relationship
+{
+  return _relationship;
+}
+@end
+
+@implementation NSStoryboardSeguePerformAction
```

Comment:
I am having trouble understanding the concept of this class. There is one set of attributes that get decode from the XIB, target, action, selector. None of these ever gets used. There is another set storyboard, storyboardsegue that get used in the code. The two variables sender and identifier are somewhat in between as they actually get used. Could you please provide me of a real live example of this class in a storyboard?

Reply:
OK, I think I understand it a bit better now. This class is no Apple class. The corresponding XML elements get created on the fly in createStoryboardProxyElementWithSelector:... that is why I could not find any real live example of that class.
It is here to represent the segue reference that was in the connection. Wouldn't it be better to actually decode this reference in GSXib5KeyedUnarchiver and treat it as another kind of connection there, like action, outlet and binding?

Reply:
Why did you mark this as resolved?

Reply:
I thought I had put an answer here, but it wasn't sent.   The reason I didn't do that, though I did consider it, was because of NSStoryboardSegue: 
1. Doesn't conform to NSCopying/NSCoding
2. Doesn't have 'target' and 'action' methods
3. This would have meant either creating a special case in NSStoryboardSegue or creating a special case in GSXib5KeyedUnarchiver.
4. Using a proxy allows for a place that the resolution of the controllers happens separately from the controller or segue code.

Reply:
You would not create an NSStoryboardSegue directly there, so 1 is invalid. What we would need to add to GSXib5KeyedUnarchiver is "segue" as another connection and a mapping from "segue" to an NSConnection subclass, which could be something like NSStoryboardSeguePerformAction but called GSStoryboardSegueConnection, which isn't much of special casing, so 3 is invalid as well. As for 2, neither does an outlet or binding connection has an action. You might have a point in 4 as the instantiation of the segue seems to happen at a different point in time, but I am really not sure here. That code is just confusing.

Reply:
Is this something we are talking about doing prior to merge?  This seems like a serious rewrite.

Reply:
Actually, the reason this can't be done that way is that the destination of this object should not be instantiated the connection is invoked.  Under normal circumstances (like with an action, outlet, binding) the  connection is made via the establishConnection method.   In the case of bindings by making a "private" subclass and invoking code before and after the accessor methods are called.   I feel better having the NSStoryboardPerformAction (though I could probably rename it) since it ensures number 4 above.

Reply:
As I wrote I am not sure whether that solution would be feasible and I must admit I did not fully understand your answer. It would definitely be worthwhile to test that approach though. At least we would need an example where this approach would not work.
Anyway, I am away for the weekend feel free to do what ever you want with this pull request.

Reply:
Just to explain the proposed approach a bit more. I think when instantiating the GSStoryboardSegueConnection (or what ever we call it) this will just result in the corresponding segue being created and stores in the controllers map (very similar to the current code). But that map would only exist on the controller. The question is still where and when doAction: (or what ever replaces it) would be called. Perhaps you could explain where this currently happens?

Reply:
okay, to break it down...

A normal connection object is created when "establishConnection" is called.  In all other connection classes this method is called unconditionally and sets the appropriate attributes to make the connection possible.  target/action in the case of an action and it associates the setter/getter (or instance var) in the case of an outlet.   In the case of a segue the destination controller is not instantiated unless the "shouldPerform" method returns YES.   This cannot happen in a traditional connection.  Therefore we cannot model a segue as a traditional connection.  This approach occurred to me when I originally did this.  In order to delay the instantiation of the destination controller I reasoned that there should be an object which acts as a go-between.

The other issue is that I didn't want to change the XIB decoding too much and figured it would be better to create an object in the XML which would be decoded on the fly and which could conform to NSCopying/NSCoding where NSStoryboardSegue could not.  Your connection class would achieve that goal, but not the other one.

Reply:
This approach merits further discussion and thought.

Reply:
I am looking at the code and thinking about this tonight.  I will report my findings later on.  I don't know if it is possible to model this purely as a connection, but I don't want to dismiss the possibility just yet.

Reply:
I think the current implementation is sufficient at this point.